api/src/main/scala/Geoprocessing.scala

```diff
@@ -19,73 +19,74 @@ trait Geoprocessing extends Utils {
-
-                case "RasterGroupedAverage" =>
-                  targetLayer match {
-                    case Some(tl) =>
-                      if (layers.isEmpty) rasterAverage(tl, shape, opts)
-                      else rasterGroupedAverage(layers, tl, shape, opts)
-                    case None =>
-                      throw new MissingTargetRasterException
-                  }
-
-                case "RasterLinesJoin" =>
-                  input.streamLines match {
-                    case Some(mls) => {
-                      val lines = (parseMultiLineString(mls) & shape).asMultiLine.get
-                      rasterLinesJoin(layers, Seq(lines)).mapValues(_.toDouble)
-                    }
-                    case None =>
-                      throw new MissingStreamLinesException
-                  }
+    val futureRasters: Map[RasterID, Future[RasterLayer]] =
+      rasterIds.map { rid =>
+        rid -> Future(cropSingleRasterToAOI(rid, 0, aoi))
+      }.toMap
+
+    val tabular: Future[List[(HucID, OperationID, Map[String,Double])]] = Future.sequence {
+      (input.shapes, input.operations).mapN { case (huc, op) =>
+        val shape = hucs(huc.id)
+        val opts = cachedOpts(op)
+
+        val futureLayers: Future[List[RasterLayer]] =
+          op.rasters.map(futureRasters(_)).sequence
+
+        val futureTargetLayer: Future[Option[RasterLayer]] =
+          op.targetRaster.map(futureRasters(_)).sequence
+
+        for {
+          layers <- futureLayers
+          targetLayer <- futureTargetLayer
+        } yield {
+          val results = op.name match {
+            case "RasterGroupedCount" =>
+              rasterGroupedCount(layers, shape, opts).mapValues(_.toDouble)
+
+            case "RasterGroupedAverage" =>
+              targetLayer match {
+                case Some(tl) =>
+                  if (layers.isEmpty) rasterAverage(tl, shape, opts)
+                  else rasterGroupedAverage(layers, tl, shape, opts)
+                case None =>
+                  throw new MissingTargetRasterException
+              }
+
+            case "RasterLinesJoin" =>
+              input.streamLines match {
+                case Some(mls) => {
+                  val lines = (parseMultiLineString(mls) & shape).asMultiLine.get
+                  rasterLinesJoin(layers, Seq(lines)).mapValues(_.toDouble)
+                }
+                case None =>
+                  throw new MissingStreamLinesException
               }
-            }
           }
+
+          (huc.id, op.label, results)
+        }
       }
     }
 
-    val result1: Map[HucID, Future[Map[OperationID, Map[String, Double]]]] =
-      result0.mapValues(liftFuture)
-
-    val result2: Future[Map[HucID, Map[OperationID, Map[String, Double]]]] =
-      liftFuture(result1)
+    val nested: Future[Map[String,Map[String,Map[String,Double]]]] = tabular.map {
+      _.groupBy { case (a, b, c) => a }.toMap.mapValues {
```

Comment:
The `.toMap` here is extraneous, since `_.groupBy` returns a map.

Reply:
d'oh. Good catch